buildSrc/src/main/groovy/org/elasticsearch/gradle/BuildPlugin.groovy

```diff
@@ -969,150 +780,127 @@ class BuildPlugin implements Plugin<Project> {
             project.tasks.withType(Test) { Test test ->
                 File testOutputDir = new File(test.reports.junitXml.getDestination(), "output")
 
-                doFirst {
+                ErrorReportingTestListener listener = new ErrorReportingTestListener(test.testLogging, testOutputDir)
+                test.extensions.add(ErrorReportingTestListener, 'errorReportingTestListener', listener)
+                test.addTestOutputListener(listener)
+                test.addTestListener(listener)
+
+                /*
+                 * We use lazy-evaluated strings in order to configure system properties whose value will not be known until
+                 * execution time (e.g. cluster port numbers). Adding these via the normal DSL doesn't work as these get treated
+                 * as task inputs and therefore Gradle attempts to snapshot them before/after task execution. This fails due
+                 * to the GStrings containing references to non-serializable objects.
+                 *
+                 * We bypass this by instead passing this system properties vi a CommandLineArgumentProvider. This has the added
+                 * side-effect that these properties are NOT treated as inputs, therefore they don't influence things like the
```

Comment:
We don't want that here, that would mean that changing from one version of java to annother ( runtime.java ) would not cause the test to be re-run as these are not treated as input.

Reply:
The runtime Java version is tracked by the `Test` task via the `bootstrapClasspath` property, not a system property. This property is simply informative and used in the reproduction line. In theory I could set `runtime.java` to "99" and the tests would care less. What _actually_ sets the runtime java version is below where we set `test.executable`.